src/bindgen/parser.rs

```diff
@@ -563,6 +572,160 @@ impl Parse {
+                self.check_nested_assoc_types(key, &mut **ty_);
+            }
+            Type::Array(ty_, _) => {
+                self.check_nested_assoc_types(key, &mut **ty_);
+            }
+            Type::FuncPtr { ret, args, .. } => {
+                self.check_nested_assoc_types(key, &mut **ret);
+                for (_, ty_) in args {
+                    self.check_nested_assoc_types(key, ty_);
+                }
+            }
+            // Base case, found a GenericPath with associated types
+            Type::Path(generic_path) => {
+                if let Some(id) = generic_path.assoc() {
+                    // Check if already resolved
+                    if let Some((ty, 0)) = self.assoc_types.get(id) {
+                        *conc_type = ty.clone();
+                    }
+                    // Unresolved nested associated types
+                    else {
+                        // Add unresolved key to pending map
+                        self.pending_assoc_types
+                            .entry(id.clone())
+                            .and_modify(|id_vec| {
+                                id_vec.push(key.clone());
+                            })
+                            .or_insert(vec![key.clone()]);
+
+                        // Increment self count
+                        self.assoc_types
+                            .entry(key.clone())
+                            .and_modify(|(_, count)| {
+                                *count += 1;
+                            });
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    // Given key k that is fully resolved, check what depends on k and resolve those
+    fn check_satisfy(&mut self, key: &AssocTypeId, resolved_ty: &Type) {
+
+        if let Some(id_vec) = self.pending_assoc_types.get(key) {
+            // Resolve for each dependent key
+            for id in id_vec {
+                let (curr_ty, count) = self.assoc_types.get_mut(id).unwrap();
+                self.check_satisfy_id(key, resolved_ty, id, curr_ty, count);
+            }
+        }
+    }
+
+    fn check_satisfy_id(&mut self, resolved_key: &AssocTypeId, resolved_ty: &Type, curr_key: &AssocTypeId, curr_ty: &mut Type, count: &mut u8) {
+        match curr_ty {
+            Type::Ptr { ty: ty_, .. } => {
+                self.check_satisfy_id(resolved_key, resolved_ty, curr_key, ty_, count);
+            }
+            Type::Array(ty_, _) => {
+                self.check_satisfy_id(resolved_key, resolved_ty, curr_key,  ty_, count);
+            },
+            Type::FuncPtr { ret, args, .. } => {
+                self.check_satisfy_id(resolved_key, resolved_ty, curr_key, ret, count);
+                for (_, ty_) in args {
+                    self.check_satisfy_id(resolved_key, resolved_ty, curr_key, ty_, count)
+                }
+            }
+            // Base case
+            Type::Path(generic_path) => {
```

Comment:
this recursive match pattern is duplicated in 3 places now. Let's be a little bit smarter about it. Let's make the following function:

```rs
impl Type {
    pub fn resolve_assoc<F: FnMut(&super::AssocTypeId) -> Option<Type>>(&mut self, fun: &mut F) {
        match self {
            Type::Ptr { ty, .. } => {
                ty.resolve_assoc(fun);
            }
            Type::Array(ty, _) => {
                ty.resolve_assoc(fun);
            }
            Type::FuncPtr { ret, args, .. } => {
                ret.resolve_assoc(fun);

                for (_, arg) in args {
                    arg.resolve_assoc(fun);
                }
            }
            Type::Path(generic_path) => {
                if let Some(assoc) = generic_path.assoc() {
                    if let Some(resolved) = fun(assoc) {
                        *self = resolved;
                    }
                }
            }
            Type::Primitive(_) => {}
        }
    }
}
```

and then you can call this function in different places